### ISSUE TITLE
Tracked Entity remains allocated until new request

### ISSUE THREAD

**Issue by czioutas** [dotnet/efcore#21092]
### Describe the bug
I did not really want to create this as a bug, because I am not sure if this is expected functionality or actually a bug.
Nevertheless my issue in short is that EF keeps the objects created (tracked) even after the request has been completed.
I register my Dbcontext with default scoped lifetime cycle and similarly the Repository that has an instance of the Dbcontext.
As I understand (could be wrong) after an http request has been completed we should no longer have such object allocations as the repository and dbcontext should be released per request/scoped lifecycle.

What I am currently seeing is illustrated bellow:

![image](https://user-images.githubusercontent.com/2229298/83346276-542e8200-a31b-11ea-8e61-b39ba31874a2.png)

In snapshot 1 or RedLine 1 my code is generating some data
In snapshot 2 or RedLine 2 my code has finished adding and saving the data to the db and EXITS
this is important, this snapshot was taken after the request returned Ok, so I would assume the allocated memory in order to add the objects and save them would be released.

In snapshot 3 or RedLine 3 (I was a bit late, wanted to take this one a bit earlier) I initiated the same exact request to see what it does on round 2 of the same call. and as we can see only after that it started freeing the memory

In snapshot 4 or RedLine 4 memory allocation is increasing normally as the code generates new data yet again.

**My problem** is that we have a processes that uses almost all of the machine's RAM but manages to finish the task fine. However the moment we try to initiate a new request, because the memory is still allocated it never reaches the "new" point in which the memory is freed as per snapshot 3 and thus dies.

I run this multiple times and tried offloading the process even to hangfire but same results.
![image](https://user-images.githubusercontent.com/2229298/83346384-2138be00-a31c-11ea-83a8-9d1882728784.png)

The above always happens, increase till request done, then continue using same memory, new request, memory freed.

From what I understand (it's my first time memory debugging) the objects are still held up due to EntityReferenceMap and I do not know if this is expected functionality.

I thought it might be Npgsql related, but I tested with in-memory db and had the same result

![image](https://user-images.githubusercontent.com/2229298/83346519-6dd0c900-a31d-11ea-884a-f56744616e36.png)

![image](https://user-images.githubusercontent.com/2229298/83346546-9c4ea400-a31d-11ea-95fd-10ef9a711726.png)

### To Reproduce
I have created a project here [https://github.com/czioutas/efcorememorytest](https://github.com/czioutas/efcorememorytest) which I also used to generated the images above.

#### IDE
Microsoft Visual Studio Community 2019
Version 16.6.0
#### EF Version
```
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.4"/>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="3.1.4"/>
    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="3.1.4"/>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.4"/>
```
#### dotnet --info
```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.300
 Commit:    b2475c1295

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.300\

Host (useful for support):
  Version: 3.1.4
  Commit:  0c2e69caa6

.NET Core SDKs installed:
  3.0.100 [C:\Program Files\dotnet\sdk]
  3.1.102 [C:\Program Files\dotnet\sdk]
  3.1.300 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by czioutas** [comment]
Made all the changes as suggested by @roji and @JeremyLikness 
- new scope,dbcontext,repository per batch 
- async/disposable changes
- batching operation, which deals with batches of 1k elements out of 100k

https://github.com/czioutas/efcorememorytest/blob/master/src/Controllers/DefaultController.cs#L70

```
using (var scope = _serviceProvider.CreateScope())
using (var _context =(ProjectDbContext)scope.ServiceProvider.GetRequiredService(typeof(ProjectDbContext)))
using (IRepository<TestModel> _repository = new DbContextRepository<TestModel>(_context))
{
        foreach (TestModel _tm in elements)
        {
             _repository.Add(_tm);
        }

        await _repository.SaveAsync(); // final save for items outside of the batch
}
```

Unfortunately there is no difference/change
I saw some more aggressive GC action but nothing freed

![image](https://user-images.githubusercontent.com/2229298/83357555-53264080-a36d-11ea-9973-b66c4fd2794b.png)

and once again memory starts getting freed when the request is started the second time (red line is when I initiated the 2nd call)
![image](https://user-images.githubusercontent.com/2229298/83357713-4d7d2a80-a36e-11ea-879b-25cbb4d77775.png)

**Comment by roji** [comment]
I haven't looked in depth, but note that you're still generating you entire data set (a list of 100000 instances of TestModel), and only then batching the DbContext operation. That means that your 100k instances stay in memory for the duration of the entire request - the batching only affects the EF Core side. Since TestModel seems to have lots of data and lots of nested entities, that is probably the source of the problem.

Try rewriting the logic to actually *produce* batches of 1000 entities at a time, so you never have more in memory. In a real life application, you'd be reading data in batches and saving it to the database, as opposed to reading the entire data into memory, and then saving it to the db batch by batch.

**Comment by czioutas** [comment]
This is very interesting. I switched to 1k batches that generate the data inside the loop and something new came up

Bellow you can see that now out of the total 100k the last 30k remain so the GC partially worked or at least better, but these are huge allocations. 30k objects out of which only the 1k should be there.

furthermore now Npgsql is doing magic by keeping lots of Npgsql Params laying around...

![image](https://user-images.githubusercontent.com/2229298/83358932-37c03300-a377-11ea-933a-d5eba3db1e5d.png)

Nevertheless at this point speed/efficiency is not the issue. The problem is the memory hog and I cannot understand where it is coming from. However I do not believe it is EF Core related anymore.
Open for input.

**Comment by roji** [comment]
@czioutas as discussed above, nothing is really being "kept around" - garbage collectors simply don't reclaim memory right away, but rather every now and then, or when there's memory pressure. There's no Npgsql magic around keeping NpgsqlParameter instances laying around - these are simply temporary gen0 instances used for the inserts, and which are discarded afterwards - we're talking about 1.3MB for the entire 30k, which is quite normal/expected (the fact that these allocations are needed at all is somewhat unfortunate, but that's another story). As you noticed, by switching to batching, you're allowing the GC to reclaim objects before the request ends, if it chooses to do that.

To sum it up, the above memory snapshot looks pretty decent to me - I can't really see any anomalies or issues. Let me know if you think otherwise.

**Comment by czioutas** [comment]
@roji sorry for the late reply, decided to actually take the monday holiday :P

I see your point and I guess it was a wrong expectation on how (and how early) the GC collects the memory. We can alter our application to fit this model. We wont have time to do this now so I guess we can close the issue until we implement on our side this smaller batching and report back in the future.

**Comment by roji** [comment]
@czioutas reopening to make 100% sure there's no issue here:

* Is your process actually running out of memory (i.e. crashing)?
* If the process is crashing, is this happening only because your request attempted to hold the entire dataset in memory (without batching), and the dataset was simply too big to hold in memory? If it's the latter, then things are working as expected.
* If the process isn't crashing, then is the problem simply that memory is only being freed *later* than expected? If so, then as discussed above things are probably working as expected.
* However, if you *are* seeing an out-of-memory crash, and you're not explicitly referencing a huge dataset in memory within a single request (i.e. no batching), then I definitely need to investigate this further.

Can you please confirm?

**Comment by pavlalexandr** [comment]
I had something similar with my console .NET Core app. The solution was to do not use Microsoft built-in DI container for DBContext lifetime. Somehow and somewhere it stores every pointer for Scoped and Transient objects, so they can't be released. 
So 

```
using(var dbContext = new MyDbContext())
{
}
```
is the best thing which can be done for now.

**Comment by czioutas** [comment]
@roji sorry for the late reply.
So maybe 1 thing I forgot to mention is that the app runs within docker.
We made the changes RE scope and today it did run into memory issues and just stopped working. The .net core inside the container just stopped and had to restart the container.

The application is an ETL so its only purpose is to do that. read files, transform and load them onto the db.
We are dealing with massive amounts of entities (300k) and a larger one is coming so we are trying to get ready but I also understand that by your side this is not normal operation for ef so I dont expect you guys to magically solve it :)

I will be moving the application outside of the container in the coming days so maybe we can remove at least that also from the "variables" that can affect it.

@pavlalexandr ye kind of what problem I have atm and switching to creating new objects out of a new ServiceScope helps. I just dont get if this is a bug or expected behavior.

**Comment by roji** [comment]
@czioutas if there's a problem here, I'm definitely interested in reproducing and understanding this even if it isn't purely an EF Core issue.

Are you enforcing a memory limit on your docker container? If so, how much is it? Did you see the crash after switching to batching, or does that occur only when loading the entire 300k entities into memory as part of the request? In any case, I'll examine your repro above in more detail and we'll see what I come up with.

**Comment by czioutas** [comment]
@roji we do not set any limits to the docker container.
We do not have batching enabled only the scoped dbcontext (batching is difficult to implement right now) - so the crash happened in an app that generates all the models (300k) then saves them in one dbcontext (although we could change and create a new dbcontext by passing the service scope).
but the crash doesnt happen from that request, it happens from the subsequent that require memory and as the previous one takes almost 90% of the ram, the new one crashes the app.

But it is not easy to reproduce and it takes a lot of time :/ 

I think I would have to make some changes to the example repo to make it as close to the current app possible and then try to crash it :P

**Comment by roji** [comment]
Thanks @czioutas, I'll take a look at this in the coming days and let you know. Regardless of any issue here, you're definitely better off doing batching rather than loading the entire dataset - this way your maximum memory requirements would be fixed.

**Comment by robi26** [comment]
I had exactly the same problem. Finally solved it (partially) by detaching the entities before disposal of the dbcontext
```csharp
var changes = dbContext.ChangeTracker.Entries<TEntity>();
foreach (var item in changes)
{
 item.State = EntityState.Detached;
}
```

**Comment by ShreyasJejurkar** [comment]
If I understood correctly above discussion, does it mean that if 1st request hits the server, it allocates a bunch of entity objects in memory to do the EF thing, once the request is finished, the DbContext gets disposed but not the entity objects? If that is so if my 2nd request comes after 10 min, for that whole time the objects are still are in memory?

And the objects get collected only after if GC feels memory pressure!? 

Consider a scenario, GC memory pressure limit is 500 MB, now if 1st request allocates and consumes 300MB memory for entity objects and the request ends, the objects are still in memory btw. And now the second request comes and somehow GC finds that the current request might allocate another 300 MB for entity objects, but now 300 MB (from the previous request) + 300 MB (will be from current request), as it exceeds GC memory pressure limit of 500 MB, now GC gets triggered in between request to clean up those previous request objects instead of serving the incoming HTTP request. Is my understanding correct? Is this what the behavior is as of now?

**Comment by roji** [comment]
@ShreyasJejurkar yes - that should be the current behavior. This has nothing to do with EF Core - that's just how garbage collection works; once the DbContext is disposed, that context and all the entity instances it referenced are now **eligible** for garbage collection (assuming no user code is referencing them). When that memory actually gets reclaimed is up to the .NET garbage collector.

So yes, if each web request allocates 300MB and the total memory limit is 500MB, then a GC pass will be triggered at the beginning of each request. This is the expected behavior.

This issue is kept open, because there seemed to have been an assertion above at some point that the process is running out of memory (crashing). That shouldn't happen, but as long as the process still runs, **when** exactly memory gets reclaimed isn't a detail that EF Core can control.

**Comment by roji** [comment]
I'm going to go ahead and close this, as there doesn't seem to be an issue here. If we get reports (or a code sample) of memory not getting released - without user code referencing it - I'll definitely reinvestigate.
